ext/node/zlib/alloc.rs

```diff
@@ -0,0 +1,62 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+
+// Workaround for https://github.com/rust-lang/libz-sys/issues/55
+// See https://github.com/rust-lang/flate2-rs/blob/31fb07820345691352aaa64f367c1e482ad9cfdc/src/ffi/c.rs#L60
+use std::alloc::Layout;
+use std::alloc::{self};
+use std::os::raw::c_void;
+use std::ptr;
+
+const ALIGN: usize = std::mem::align_of::<usize>();
+
+fn align_up(size: usize, align: usize) -> usize {
+  (size + align - 1) & !(align - 1)
+}
+
+pub extern "C" fn zalloc(
+  _ptr: *mut c_void,
+  items: u32,
+  item_size: u32,
+) -> *mut c_void {
+  // We need to multiply `items` and `item_size` to get the actual desired
```

Comment:
All of this magic already exists in `Box::`[into_raw](https://doc.rust-lang.org/std/boxed/struct.Box.html#method.into_raw) which I think can simplify this quite a bit.

Reply:
How would you use Box::into_raw here? There is no concrete type we know; this is an allocator impl for zlib internals so that it runs through the same global allocator in Rust.

Reply:
I didn't see a specific alignment guarantee required for the zalloc function, but I included both aligned and unaligned allocators that are more "rust-y" and automatically manage everything. If the performance of a zeroed allocator has measurable impact, there are other options.

These compile down to a raw allocation under the hood, and uses the Rust allocator as you'd expect with no surprises:

```
#[inline]
pub fn unaligned_alloc(len: usize) -> *mut u8 {
    let v = vec![0; len];
    let ptr = Box::into_raw(v.into_boxed_slice());
    ptr as *mut u8
}

pub fn unaligned_dealloc(ptr: *mut u8) {
    let b = unsafe { Box::from_raw(ptr) };
    drop(b);
}

/// Note that len is in units of A's size, not bytes.
#[inline]
pub fn aligned_alloc<A: Copy + Default>(len: usize) -> *mut A {
    let v = vec![A::default(); len];
    let ptr = Box::into_raw(v.into_boxed_slice());
    ptr as *mut A
}

pub fn aligned_dealloc<A: Copy + Default>(ptr: *mut A) {
    let b = unsafe { Box::from_raw(ptr) };
    drop(b);
}

pub fn main() {
    let len = 1024;
    let ptr = unaligned_alloc(len);
    unaligned_dealloc(ptr);
    
    let ptr = aligned_alloc::<u64>(len);
    aligned_dealloc(ptr);
}

```

It compiles down to a zeroed allocation, aligned as you'd expect:

https://godbolt.org/z/T1xob8Tf6

Reply:
Here's the thing: in `zfree` we don't know the size of the allocation. In your `unaligned_dealloc`, `Box::from_raw(ptr as *mut u8)` is `Box<u8>`. - you need a fat pointer though i.e. call `slice::from_raw_parts` with the size.

The reason I chose this implementation is because it's done in a popular Rust crate (flate2): https://github.com/rust-lang/flate2-rs/blob/main/src/ffi/c.rs#L68

Reply:
If you set len to 0 for your aligned and unaligned code, it will crash trying to deallocate the null pointer. 

https://rust.godbolt.org/z/zjoY9hz5q

Reply:
I didn't realize that was from `flate2`, I'm good with that if the pattern is established. I'll resolve this and check my assumptions on the box approach for later -- you might be right here.

Reply:
I think you're right here - and I think it's because the compiler neglected to warn me that I was casting from *mut [A] (the fat pointer) to *mut A (the thin pointer). So if I tried to use the pointer that came back from it I'd end up stomping over the fat pointer and beyond (as this is a pointer to the fat pointer's allocation, not the slice buffer). Good call!